File: benchmarks/matty_suite/bench_filter_accuracy.py
"""bench_filter_accuracy.py -- numpy/scipy port of bench_filter_accuracy.qu.

Same exact closed-form ground truth, same explicit frequency axis passed
to sosfreqz rather than trusted to a default (scipy's default integer
worN excludes the Nyquist endpoint -- verified directly before writing
this: sosfreqz(sos, worN=5) stops short of pi, sosfreqz(sos, worN=array)
does not. Passing linspace(0, pi, n_pts) explicitly is what makes this
comparable point-for-point against Qu's freqz(filt, n), which always
includes both 0 and pi.
"""

import time

import numpy as np
from scipy import signal

Fs = 1000.0
fc = 100.0
n_pts = 513
wc = 2.0 * np.pi * fc / Fs
orders = [2, 4, 6, 8, 10, 12, 16, 20]
w = np.linspace(0.0, np.pi, n_pts)

print("order  max_err       rms_err       design+freqz time")
for N in orders:
    t0 = time.perf_counter()
    sos = signal.butter(N, fc, btype="low", fs=Fs, output="sos")
    _, H = signal.sosfreqz(sos, worN=w)
    t = time.perf_counter() - t0

    mag = np.abs(H)
    exact_mag = np.ones_like(w)
    nz = w > 0
    ratio = np.tan(w[nz] / 2.0) / np.tan(wc / 2.0)
    # High order + w near pi sends ratio**(2N) to +inf in float64 -- the
    # limit 1/(1+inf) = 0 is still the mathematically correct answer, so
    # this is silenced rather than avoided.
    with np.errstate(over="ignore"):
        exact_mag[nz] = np.sqrt(1.0 / (1.0 + ratio ** (2 * N)))

    err = np.abs(mag - exact_mag)
    max_err = np.max(err)
    rms_err = np.sqrt(np.mean(err ** 2))

    print(f"{N:2d}     {max_err:.6e}  {rms_err:.6e}  {t:.6f} s")

File: benchmarks/matty_suite/bench_filter_accuracy.m
% bench_filter_accuracy.m -- MATLAB port of bench_filter_accuracy.qu.
%
% Same closed-form ground truth, same explicit frequency axis (0:pi,
% n_pts points, passed directly to freqz) so every language evaluates
% the identical set of frequencies -- freqz's own default point count
% does not guarantee the same axis another language's default would.
%
% UNTESTED on this machine: R2025b (the only licensed, working install --
% R2026a's own license is separately expired) has no Signal Processing
% Toolbox (checked via `ver`; only MATLAB + Parallel Computing Toolbox are
% licensed), and both `butter` and `freqz` require it. Written to match
% bench_filter_accuracy.qu/.py exactly so it is ready the moment a
% licensed install is available -- do not report a result from this file
% without actually running it first.

Fs = 1000.0;
fc = 100.0;
n_pts = 513;
wc = 2.0 * pi * fc / Fs;
orders = [2, 4, 6, 8, 10, 12, 16, 20];
w = linspace(0, pi, n_pts)';

fprintf('order  max_err       rms_err       design+freqz time\n');
for N = orders
    tic;
    sos = butter(N, fc / (Fs / 2), 'low');
    H = freqz(sos, w);
    t = toc;

    mag = abs(H);
    exact_mag = ones(n_pts, 1);
    nz = w > 0;
    ratio = tan(w(nz) / 2.0) / tan(wc / 2.0);
    exact_mag(nz) = sqrt(1.0 ./ (1.0 + ratio .^ (2 * N)));

    err = abs(mag - exact_mag);
    max_err = max(err);
    rms_err = sqrt(mean(err .^ 2));

    fprintf('%2d     %.6e  %.6e  %.6f s\n', N, max_err, rms_err, t);
end
